esmvalcore/cmor/_fixes/cmip6/giss_e2_1_g.py

```diff
@@ -9,3 +10,27 @@
 
 
 Clw = ClFixHybridPressureCoord
+
+
+class Nbp(Fix):
+    """Fixes for nbp."""
+
+    def fix_data(self, cube):
+        """Fix data.
+
+        Fixes wrong sign for land surface flux.
+
+        Parameters
+        ----------
+        cube : iris.cube.Cube
+            Input cube.
+
+        Returns
+        -------
+        iris.cube.Cube
+
+        """
+        metadata = cube.metadata
+        cube *= -1
```

Comment:
This fix is not idempotent, i.e. repeatedly applying this fix will always flip the sign. This is particularly problematic since it will unfix data if ever an updated version of it is put on ESGF.

I wonder if [`np.abs()`](https://numpy.org/doc/stable/reference/generated/numpy.absolute.html) could be used?
I also think that maybe for `fix_data` it is better to use `cube.data = np.abs(cube.core_data())`, though I am not sure. @valeriupredoi, do you know?

Reply:
aren't most of our fixes idempotent ie fixing twice is not good? (BTW awesome word, had to look it up :grin: )

Reply:
Negative fluxes are possible and occuring for this variable, as the land acts as both a sink and a source for carbon depending on region and time, so using the absolute value is not possible. As I said the GISS fix is particularly weird at the moment and I'd leave it out, so the discussion would be for how to deal with it for SAM0-UNICON.

Reply:
> aren't most of our fixes idempotent ie fixing twice is not good? (BTW awesome word, had to look it up grin )

Just to be clear: If a fix is idempotent, it means that you can apply it as many times as you want; only the first time will actually change anything.

Since we don't check versions on the data, and since updated and fixed data gets published to ESGF all the time, this is pretty much a necessary property. If existing fixes have it is a bit besides the point.

Anyway, any knowledge on the lazyness of data fixes and the use of `.core_data` here?

Reply:
> Negative fluxes are possible and occuring for this variable, as the land acts as both a sink and a source for carbon depending on region and time, so using the absolute value is not possible. As I said the GISS fix is particularly weird at the moment and I'd leave it out, so the discussion would be for how to deal with it for SAM0-UNICON.

Right. I just commented on the GISS version because the code lives here and is then used verbatim for the other case as well. Is there any chance to detect whether it's right or wrong? Perhaps with a heuristic, say over a large forest area it should be predominantly positive? If this is all too much effort I suggest to at least leave a comment, perhaps mentioning the version that has the fault.

Reply:
sorry - was writing a reply and then I got distracted - `core_data()` is a dask array so we should flip its sign with a dask method, unfortunately I looked around (hence my distraction and not posting the reply) but I couldn't find any - dask doesn't even have a builtin `abs` afaik

Reply:
> Is there any chance to detect whether it's right or wrong? Perhaps with a heuristic, say over a large forest area it should be predominantly positive? If this is all too much effort I suggest to at least leave a comment, perhaps mentioning the version that has the fault.
Due to the seasonal cycle you'd have to fix not just an area but also a time to check for an overall sign. This may also not hold for some scenarios that could be explored. Adding the version that needed the fix as a comment is a better idea, detecting it as an absolute for all experiments and times seems too hard.

Reply:
> > Is there any chance to detect whether it's right or wrong? Perhaps with a heuristic, say over a large forest area it should be predominantly positive? If this is all too much effort I suggest to at least leave a comment, perhaps mentioning the version that has the fault.
> 
> Due to the seasonal cycle you'd have to fix not just an area but also a time to check for an overall sign. This may also not hold for some scenarios that could be explored. Adding the version that needed the fix as a comment is a better idea, detecting it as an absolute for all experiments and times seems too hard.

I thought that might be the case and understand. Adding a comment with the version would be great.

Reply:
> sorry - was writing a reply and then I got distracted - `core_data()` is a dask array so we should flip its sign with a dask method, unfortunately I looked around (hence my distraction and not posting the reply) but I couldn't find any - dask doesn't even have a builtin `abs` afaik

Well, it's like this: `core_data()` can be either a dask array or a numpy array, so we need something that works with both. Luckily, numpy as so-called dispatching, that is, if you apply a numpy operation to a non-numpy object, that object can tell numpy what to do. This works for many numpy functions, but in particular for ufuncs. Here, `np.abs` is an alias for `np.absolute`, which is a ufunc. This is mirrored by `da.absolute` which exists, but doesn't show up in the api docs because as a ufunc it has been integrated automatically.

Practical upshot: `cube.data = np.abs(cube.core_data())` will put either a numpy or a dask array into `cube.data`, depending on the initial type of `.core_data()` that has been treated with the correct numpy or dask version of `absolute`, keeping the data lazy in case it was lazy before. Of course, for `cube.data = - cube.core_data()` essentially the same applies.

`cube.data *= -1` on the other hand realizes the data needlessly, if I understand correctly.

Reply:
nops, the PR has `cube *= -1` in it no? That don't realize the data, keeps it lazy

Reply:
also `core_data()` is always a dask array when data is lazy, applying a numpy abs on it keeps it lazy so I am OK with `np.abs`-ing `core_data()` if that's needed, but I thought all we had to do here was multiply by -1 to flip the sign of le data?

Reply:
Right. So in fixes when using functions and stuff we should use core_data, but here things work out and all we need to do is add the comment discussed above.

Reply:
yepers dapers :grin:

Reply:
@LisaBock would be able to add the comment @zklaus is suggesting please, then I or Remi can merge this - good stuff :+1:

Reply:
Lisa's on holidays, so I did it now. Also removed the GISS fix since it's not universal for every ensemble member. Can you double check that I didn't introduce errors in the tests by copy pasting the stuff over?